### PR TITLE
[#10201] Update priority for ECDSA KEX.

### PR DESCRIPTION
**Pull request by adiroiban**
## Scope and purpose

I think that ECDSA 512  should have higher priority than ECDSA 256.

In conch _kex the lower priority is the preferred algorithm.

## Contributor Checklist:

* [x] The associated ticket in Trac is here: https://twistedmatrix.com/trac/ticket/10201
* [x] I ran `tox -e lint` to format my patch to meet the [Twisted Coding Standard](https://twistedmatrix.com/documents/current/core/development/policy/coding-standard.html)
* [x] I have created a newsfragment in src/twisted/newsfragments/ (see: [News files](https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles))
* [x] The title of the PR starts with the associated Trac ticket number (without the `#` character).
* [x] I have updated the automated tests and checked that all checks for the PR are green.
* [x] I have submitted the associated Trac ticket for review by adding the word `review` to the keywords field in Trac, and putting a link to this PR in the comment; it shows up in https://twisted.reviews/ now.
* [x] The merge commit will use the below format
  The first line is automatically generated by GitHub based on PR ID and branch name.
  The other lines generated by GitHub should be replaced.

```
Merge pull request #123 from twisted/4356-branch-name-with-trac-id

Author:  adiroiban
Reviewer: alex
Fixes: ticket:10201

Document that ecdh-sha2-nistp521 has lower priority than ecdh-sha2-nistp256 on purpose.
```

### CODE REVIEW

**Comment by cjwatson** [comment]
Before going ahead with this, it may be worth checking with the OpenSSH developers why they elected to put `ecdh-sha2-nistp256` first in their equivalent preference order.  I had a look in git history there and couldn't see an obvious rationale, but there may have been one.

**Comment by glyph** [comment]
@alex or @reaperhulk - could I bug one of you to get a quick rationale for this priority change, if you happen to know?

**Comment by alex** [comment]
No.

There is not a good reason to use P-521 over P-256. There's no meaningful security improvement, there's a huge performance degredation, and basically all wiely used P-521 impls are not constant time.

If you want to promote a different curve, promote X25519 or x448. P-521 is bad news.

**Comment by adiroiban** [comment]
Thanks for the feedback.

X25519 is already promoted as the to priority.

I will revert the change and add a comment about why this priority was used.

I will keep/update the tests as I think that we don't have any functional tests for the priority list presented to the peer SSH party..


AFAIK OpenSSH does not yet support X448.

```
adi@g14:~$ ssh -V
OpenSSH_8.4p1 Ubuntu-5ubuntu1, OpenSSL 1.1.1j  16 Feb 2021


adi@g14:~$ ssh -Q kex
diffie-hellman-group1-sha1
diffie-hellman-group14-sha1
diffie-hellman-group14-sha256
diffie-hellman-group16-sha512
diffie-hellman-group18-sha512
diffie-hellman-group-exchange-sha1
diffie-hellman-group-exchange-sha256
ecdh-sha2-nistp256
ecdh-sha2-nistp384
ecdh-sha2-nistp521
curve25519-sha256
curve25519-sha256@libssh.org
sntrup4591761x25519-sha512@tinyssh.org


adi@g14:~$ ssh -Q key
ssh-ed25519
ssh-ed25519-cert-v01@openssh.com
sk-ssh-ed25519@openssh.com
sk-ssh-ed25519-cert-v01@openssh.com
ssh-rsa
ssh-dss
ecdsa-sha2-nistp256
ecdsa-sha2-nistp384
ecdsa-sha2-nistp521
sk-ecdsa-sha2-nistp256@openssh.com
ssh-rsa-cert-v01@openssh.com
ssh-dss-cert-v01@openssh.com
ecdsa-sha2-nistp256-cert-v01@openssh.com
ecdsa-sha2-nistp384-cert-v01@openssh.com
ecdsa-sha2-nistp521-cert-v01@openssh.com
sk-ecdsa-sha2-nistp256-cert-v01@openssh.com
```

**Review by alex — APPROVED** [review]

**Comment by adiroiban** [comment]
Thanks Alex for the review.

I have reverted the implementation 

In RFC 5656 there is this table https://datatracker.ietf.org/doc/html/rfc5656#section-1

```
      +-----------+------------------------------+-------+---------+
      | Symmetric | Discrete Log (e.g., DSA, DH) |  RSA  |   ECC   |
      +-----------+------------------------------+-------+---------+
      |     80    |       L = 1024, N = 160      |  1024 | 160-223 |
      |           |                              |       |         |
      |    112    |       L = 2048, N = 256      |  2048 | 224-255 |
      |           |                              |       |         |
      |    128    |       L = 3072, N = 256      |  3072 | 256-383 |
      |           |                              |       |         |
      |    192    |       L = 7680, N = 384      |  7680 | 384-511 |
      |           |                              |       |         |
      |    256    |      L = 15360, N = 512      | 15360 |   512+  |
      +-----------+------------------------------+-------+---------+
```


In RFC 7748  https://datatracker.ietf.org/doc/rfc7748/ I see that Curve25519 has about 128 bit security level

--------

My understanding is that Curve25519 is just faster and without NIST influence , but not "stronger" than ECDSA.

--------

but if you really want a specific method in Twisted there is always the option to set up custom priorities.

This PR is for the default priorities.

File: src/twisted/conch/ssh/_kex.py [hunk 1]
```diff
@@ -92,6 +92,12 @@ class _ECDH256:
     """
     Elliptic Curve Key Exchange with SHA-256 as HASH. Defined in
     RFC 5656.
+
+    Note that C{ecdh-sha2-nistp256} takes priority over nistp384 or nistp512.
+    This is the same priority from OpenSSH.
+
+    C{ecdh-sha2-nistp256} is considered preety good cryptography.
```
Comment (graingert) [inline src/twisted/conch/ssh/_kex.py:99]:
```suggestion
    C{ecdh-sha2-nistp256} is considered pretty good cryptography.
```

by who? is this a Pretty Good Privacy (PGP) reference?

**Review by graingert — APPROVED** [review]

File: src/twisted/conch/ssh/_kex.py (hunk 1 above)
Comment (adiroiban) [reply src/twisted/conch/ssh/_kex.py:99]: sorry. I will fix the typo as a drive by in another PR.

by who?  Alex https://github.com/twisted/twisted/pull/1602#issuecomment-869056649

thanks

**Pull request merged**